incubator/hnc/internal/forest/namespaceobjects.go

```diff
@@ -5,80 +5,76 @@ import (
 }
 
-// GetOriginalObject gets an original object by name. It returns nil, if the object doesn't exist.
-func (ns *Namespace) GetOriginalObject(gvk schema.GroupVersionKind, nm string) *unstructured.Unstructured {
-	return ns.originalObjects[gvk][nm]
+// GetSourceObject gets a source object by name. If it doesn't exist, return nil.
+func (ns *Namespace) GetSourceObject(gvk schema.GroupVersionKind, nm string) *unstructured.Unstructured {
+	return ns.sourceObjects[gvk][nm]
 }
 
-// HasOriginalObject returns if the namespace has an original object.
-func (ns *Namespace) HasOriginalObject(gvk schema.GroupVersionKind, oo string) bool {
-	return ns.GetOriginalObject(gvk, oo) != nil
+// HasSourceObject returns if the namespace has a source object.
+func (ns *Namespace) HasSourceObject(gvk schema.GroupVersionKind, oo string) bool {
+	return ns.GetSourceObject(gvk, oo) != nil
 }
 
-// DeleteOriginalObject deletes an original object by name.
-func (ns *Namespace) DeleteOriginalObject(gvk schema.GroupVersionKind, nm string) {
-	delete(ns.originalObjects[gvk], nm)
+// DeleteSourceObject deletes a source object by name.
+func (ns *Namespace) DeleteSourceObject(gvk schema.GroupVersionKind, nm string) {
+	delete(ns.sourceObjects[gvk], nm)
 	// Garbage collection
-	if len(ns.originalObjects[gvk]) == 0 {
-		delete(ns.originalObjects, gvk)
+	if len(ns.sourceObjects[gvk]) == 0 {
+		delete(ns.sourceObjects, gvk)
 	}
 }
 
-// GetOriginalObjects returns all original objects in the namespace.
-func (ns *Namespace) GetOriginalObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
+// GetSourceObjects returns all source objects in the namespace.
+func (ns *Namespace) GetSourceObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
 	o := []*unstructured.Unstructured{}
-	for _, obj := range ns.originalObjects[gvk] {
+	for _, obj := range ns.sourceObjects[gvk] {
 		o = append(o, obj)
 	}
 	return o
 }
 
-// GetNumOriginalObjects returns the total number of original objects of a specific GVK in the namespace.
-func (ns *Namespace) GetNumOriginalObjects(gvk schema.GroupVersionKind) int {
-	return len(ns.originalObjects[gvk])
+// GetNumSourceObjects returns the total number of source objects of a specific
+// GVK in the namespace.
+func (ns *Namespace) GetNumSourceObjects(gvk schema.GroupVersionKind) int {
+	return len(ns.sourceObjects[gvk])
 }
 
-// GetPropagatedObjects returns all original copies in the ancestors.
-func (ns *Namespace) GetPropagatedObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
-	o := []*unstructured.Unstructured{}
+// GetAncestorSourceObjects returns all source objects with the specified name
+// in the ancestors (including itself) from top down. If the name is not
+// specified, all the source objects in the ancestors will be returned.
+func (ns *Namespace) GetAncestorSourceObjects(gvk schema.GroupVersionKind, name string) []*unstructured.Unstructured {
+	if ns == nil {
+		return nil
+	}
```

Comment:
Maybe comment why we have this here - eg because it allows you to say `ns.GetParent().GetAncestorSourceObjects()` without caring if the namespace is a root.